File: tools/regulatory_tss.py
#!/usr/bin/env python3
"""Publish the pinned Tan 2018 non-gTSS table for the browser."""

from __future__ import annotations

import argparse
import csv
import hashlib
import json
import math
from collections import Counter
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
SOURCE = ROOT / "data/expression/tan2018_utex2973_regulatory_tss_table_s1.tsv"
TARGETS = ROOT / "data/expression/tan2018_utex2973_asrna_potential_targets_table_s8.tsv"
OUTPUT = ROOT / "site/data/regulatory_tss.json"
SOURCE_SHA256 = "98a19729bf47940bf6832e3f08c3548fc4ddc7bb866176ab70e9c92c7c8d3dbe"
TARGETS_SHA256 = "0aa810d75f1e92200044ac445a7ba4ee90652a9d82d575040a803c9d66607a69"
WORKBOOK_SHA256 = "098ecbd204cd1042a6edee1d2a500eaeca4efe034c503e2ade3db1c605e79b00"
TYPE_COUNTS = {"aTSS": 1380, "iTSS": 724, "nTSS": 229}
MAPPING_COUNTS = {"mapped": 2068, "unassociated": 180, "unmapped": 85}


def build(source: Path = SOURCE, targets: Path = TARGETS) -> bytes:
    """Validate the derived table and encode every row without numeric coercion."""
    source_bytes = source.read_bytes()
    if hashlib.sha256(source_bytes).hexdigest() != SOURCE_SHA256:
        raise ValueError("Tan regulatory TSV checksum mismatch")
    with source.open(encoding="utf-8", newline="") as handle:
        rows = list(csv.DictReader(handle, delimiter="\t"))
    if Counter(row["type"] for row in rows) != TYPE_COUNTS:
        raise ValueError("Tan regulatory TSS type counts changed")
    if Counter(row["mapping_status"] for row in rows) != MAPPING_COUNTS:
        raise ValueError("Tan regulatory TSS mapping counts changed")
    if len({row["tss_id"] for row in rows}) != len(rows):
        raise ValueError("Tan regulatory TSS identifiers are not unique")
    target_bytes = targets.read_bytes()
    if hashlib.sha256(target_bytes).hexdigest() != TARGETS_SHA256:
        raise ValueError("Tan Table S8 target TSV checksum mismatch")
    with targets.open(encoding="utf-8", newline="") as handle:
        potential_targets = list(csv.DictReader(handle, delimiter="\t"))
    if Counter(row["comparison"] for row in potential_targets) != {
        "dark": 77, "high_light": 21, "high_temperature": 3,
    }:
        raise ValueError("Tan Table S8 comparison counts changed")
    by_site = {row["tss_id"]: row for row in rows}
    discrepancies = []
    for claim in potential_targets:
        site = by_site.get(claim["tss_id"])
        if site is None or site["type"] != "aTSS" \
                or claim["potential_target_locus"] != site["source_locus_tag"] \
                or float(claim["source_selection_min_abs_log2fc"]) != {
                    "dark": 1.5, "high_light": 1.5, "high_temperature": 1.0,
                }[claim["comparison"]]:
            raise ValueError("Tan Table S8 potential target is not a source-linked aTSS")
        comparison = claim["comparison"]
        table_s1_value = site[f"{comparison}_log2fc"]
        table_s8_value = claim["atss_log2fc"]
        if not math.isclose(float(table_s1_value), float(table_s8_value), rel_tol=1e-12):
            discrepancies.append({
                "tssId": claim["tss_id"],
                "comparison": comparison,
                "tableS1Log2Fc": table_s1_value,
                "tableS8Log2Fc": table_s8_value,
            })
    if {(item["tssId"], item["comparison"]) for item in discrepancies} != {
        ("aTSS-320358", "dark")
    }:
        raise ValueError("Tan Table S1/S8 antisense discrepancy set changed")
    opposite_site = by_site.get("iTSS+320358")
    if opposite_site is None or {
        key: opposite_site[key] for key in (
            "strand", "position", "source_locus_tag", "control_1", "control_2",
            "dark_1", "dark_2", "dark_log2fc",
        )
    } != {
        "strand": "+", "position": "320358", "source_locus_tag": "M744_RS01695",
        "control_1": "0", "control_2": "9", "dark_1": "420", "dark_2": "422",
        "dark_log2fc": "-4.78047469754605",
    }:
        raise ValueError("Tan opposite-strand source caution changed")
    warnings = [
        {
            "tssId": "aTSS-320358",
            "comparison": "dark",
            "message": "Table S1 reports +8.619 log2FC, while Table S8 reports −4.780. "
                       "The Table S8 potential-target claim depends on its disputed value. "
                       "Both published results remain uncorrected.",
        },
        {
            "tssId": "iTSS+320358",
            "comparison": "dark",
            "message": "Table S1 reports −4.780 log2FC, while this site's raw dark "
                       "counts (420, 422) exceed its control counts (0, 9). "
                       "The source's value and counts remain uncorrected.",
        },
    ]
    document = {
        "schemaVersion": 1,
        "source": {
            "doi": "10.1186/s13068-018-1215-8",
            "strain": "Synechococcus elongatus UTEX 2973",
            "assay": "dRNA-seq transcription start sites",
            "workbookSha256": WORKBOOK_SHA256,
            "derivedTsvSha256": SOURCE_SHA256,
            "potentialTargetsTsvSha256": TARGETS_SHA256,
            "license": "CC BY 4.0",
            "conditions": {
                "control": "33 °C, 50 µmol photons m⁻² s⁻¹ continuous light, 3% CO₂",
                "dark": "Dark for 2 h",
                "high_light": "1,000 µmol photons m⁻² s⁻¹; duration differs between paper sections",
                "high_temperature": "45 °C for 30 min",
            },
        },
        "rows": rows,
        "potentialTargets": potential_targets,
        "sourceDiscrepancies": discrepancies,
        "sourceWarnings": warnings,
    }
    return (json.dumps(document, ensure_ascii=False, separators=(",", ":")) + "\n").encode()


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("mode", choices=("write", "check"))
    arguments = parser.parse_args()
    expected = build()
    if arguments.mode == "write":
        OUTPUT.write_bytes(expected)
    elif OUTPUT.read_bytes() != expected:
        raise ValueError("site/data/regulatory_tss.json is out of date")


if __name__ == "__main__":
    main()

File: site/js/core/regulatory-tss.js
/** Tan 2018 regulatory start sites remain independent feature records. */
export const REGULATORY_TYPES = Object.freeze([
  { id: 'all', label: 'All site types' },
  { id: 'aTSS', label: 'Antisense TSS' },
  { id: 'iTSS', label: 'Internal TSS' },
  { id: 'nTSS', label: 'Orphan or novel TSS' },
]);

export function validateRegulatoryTss(document, genes) {
  if (document?.schemaVersion !== 1 || !Array.isArray(document.rows)
    || !Array.isArray(document.potentialTargets)
    || !Array.isArray(document.sourceDiscrepancies)
    || !Array.isArray(document.sourceWarnings)
    || document.source?.derivedTsvSha256 !== '98a19729bf47940bf6832e3f08c3548fc4ddc7bb866176ab70e9c92c7c8d3dbe'
    || document.source?.potentialTargetsTsvSha256 !== '0aa810d75f1e92200044ac445a7ba4ee90652a9d82d575040a803c9d66607a69') {
    throw new Error('regulatory_tss.json has an unsupported or unpinned schema');
  }
  const loci = new Set(genes.map((gene) => gene.id));
  const counts = { aTSS: 0, iTSS: 0, nTSS: 0 };
  const mapped = new Set();
  const ids = new Set();
  for (const row of document.rows) {
    if (!Object.hasOwn(counts, row.type) || ids.has(row.tss_id)
      || !Number.isInteger(Number(row.position))
      || !['+', '-'].includes(row.strand)) {
      throw new Error('regulatory_tss.json has an invalid site record');
    }
    ids.add(row.tss_id);
    counts[row.type] += 1;
    if (row.mapping_status === 'mapped') {
      if (!loci.has(row.mapped_locus_tag)
        || row.source_locus_tag !== row.mapped_locus_tag) {
        throw new Error('regulatory_tss.json has an invalid exact-locus link');
      }
      mapped.add(row.tss_id);
    } else if (!['unassociated', 'unmapped'].includes(row.mapping_status)
      || row.mapped_locus_tag !== ''
      || (row.mapping_status === 'unassociated' && row.source_locus_tag !== '')
      || (row.mapping_status === 'unmapped'
        && (!row.source_locus_tag || loci.has(row.source_locus_tag)))) {
      throw new Error('regulatory_tss.json has an invalid mapping status');
    }
  }
  if (counts.aTSS !== 1380 || counts.iTSS !== 724 || counts.nTSS !== 229
    || mapped.size !== 2068) {
    throw new Error('regulatory_tss.json does not reconcile to Tan Table S1');
  }
  const byId = new Map(document.rows.map((row) => [row.tss_id, row]));
  const targetCounts = { dark: 0, high_light: 0, high_temperature: 0 };
  for (const claim of document.potentialTargets) {
    const site = byId.get(claim.tss_id);
    if (!Object.hasOwn(targetCounts, claim.comparison)
      || site?.type !== 'aTSS'
      || claim.potential_target_locus !== site.source_locus_tag
      || claim.atss_position !== site.position
      || Number(claim.source_selection_min_abs_log2fc)
        !== (claim.comparison === 'high_temperature' ? 1 : 1.5)) {
      throw new Error('regulatory_tss.json has an invalid potential-target claim');
    }
    targetCounts[claim.comparison] += 1;
  }
  if (targetCounts.dark !== 77 || targetCounts.high_light !== 21
    || targetCounts.high_temperature !== 3) {
    throw new Error('regulatory_tss.json does not reconcile to Tan Table S8');
  }
  if (document.sourceDiscrepancies.length !== 1
    || document.sourceDiscrepancies[0].tssId !== 'aTSS-320358'
    || document.sourceDiscrepancies[0].comparison !== 'dark') {
    throw new Error('regulatory_tss.json has an unexpected Table S1/S8 disagreement');
  }
  if (document.sourceWarnings.length !== 2
    || document.sourceWarnings[0].tssId !== 'aTSS-320358'
    || document.sourceWarnings[1].tssId !== 'iTSS+320358'
    || document.sourceWarnings.some((warning) =>
      warning.comparison !== 'dark' || typeof warning.message !== 'string'
      || !byId.has(warning.tssId))) {
    throw new Error('regulatory_tss.json has unexpected source warnings');
  }
  const counterpart = byId.get('iTSS+320358');
  if (counterpart.dark_log2fc !== '-4.78047469754605'
    || counterpart.control_1 !== '0' || counterpart.control_2 !== '9'
    || counterpart.dark_1 !== '420' || counterpart.dark_2 !== '422') {
    throw new Error('regulatory_tss.json source warning has changed evidence');
  }
  const discrepancy = document.sourceDiscrepancies[0];
  if (discrepancy.tableS1Log2Fc !== byId.get(discrepancy.tssId).dark_log2fc
    || !document.potentialTargets.some((claim) => claim.tss_id === discrepancy.tssId
      && claim.comparison === discrepancy.comparison
      && claim.atss_log2fc === discrepancy.tableS8Log2Fc)) {
    throw new Error('regulatory_tss.json source disagreement has changed evidence');
  }
  return document;
}

export function potentialTargetsBySite(claims) {
  const grouped = new Map();
  for (const claim of claims) {
    if (!grouped.has(claim.tss_id)) grouped.set(claim.tss_id, []);
    grouped.get(claim.tss_id).push(claim);
  }
  return grouped;
}

export function searchRegulatoryTss(rows, {
  query = '', type = 'all', mapping = 'all', potential = 'any', targetSiteIds = new Set(),
} = {}) {
  const needle = query.trim().toLowerCase();
  return rows.filter((row) => {
    if (type !== 'all' && row.type !== type) return false;
    if (mapping !== 'all' && row.mapping_status !== mapping) return false;
    if (potential === 'only' && !targetSiteIds.has(row.tss_id)) return false;
    if (!needle) return true;
    return [row.tss_id, row.source_locus_tag, row.mapped_locus_tag,
      row.replicon, row.position].some((value) => String(value).toLowerCase().includes(needle));
  });
}
